spec/support/matchers/have_annotations.rb

```diff
@@ -0,0 +1,43 @@
+RSpec::Matchers.define :have_annotations do |*expected|
+  actual_annotations = []
+
+  match do |actual|
+    actual_annotations = actual.annotations.map do |annotation|
+      {
+        prefix: annotation.key_prefix,
+        key: annotation.key_name,
+        value: annotation.value,
+      }
+    end
+
+    if expected.any?
+      expected = expected.map do |annotation|
+        {
+          prefix: annotation.with_indifferent_access[:prefix],
+          key: annotation.with_indifferent_access[:key],
+          value: annotation.with_indifferent_access[:value],
+        }
+      end
+
+      expect(actual_annotations).to match_array(expected)
+    else
+      expect(actual_annotations).not_to be_empty
+    end
+  end
+
+  failure_message do
+    if expected.any?
+      "Annotations don't match: Expected #{expected} got #{actual_annotations}"
+    else
+      "Annotations not expected: got #{actual_annotations}"
+    end
+  end
+
+  failure_message_when_negated do
+    if expected.any?
+      "Annotations unexpectedly match: got #{actual_annotations}"
+    else
+      "Annotations found when not expected: got #{actual_annotations}"
```

Comment:
not have_annotations([]) seems like the message should actually be "Expected annotations but got []" if I am understanding correctly the usage of these method in the matchers design.

Reply:
Thanks for the explanation on the usage of any? on the matcher context. The message is fine then.